### PR TITLE
[AIRFLOW-3036] Updates for MYSQL migrations

### PR DESCRIPTION
**Pull request by dossett**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-3036/) 

### Description

- [X] We recently upgraded from 1.9.0 to 1.10.3 (running in GCP and backed by a mysql flavor of CloudSQL) and this patch allowed out database upgrade to proceed.  Please see the JIRA ticket for discussion and links to further discussion on mailing lists.

About 75% of this patch was from someone on the mailing list and we added the additional 25% to get it working end to end.

### CODE REVIEW

**Comment by dossett** [comment]
Here's the travis error, which I think I understand.  I'll try to take a look soon, but if anyone else who is more familiar with alchemy wants to look, that would be much appreciated.

```
[0m   Traceback (most recent call last):[0m
    [34mtests/utils/test_db.py[0m line [1m[36m98[0m[0m in [36mtest_database_schema_and_sqlalchemy_model_are_in_sync[0m
      'Database schema and SQLAlchemy model are not in sync: ' + str(diff)
[31m   [31m[1m[31mAssertionError[0m[0m[31m: [0m[31m[[('modify_nullable', None, 'task_fail', 'start_date', {'existing_type': TIMESTAMP(fsp=6), 'existing_server_default': DefaultClause(<sqlalchemy.sql.elements.TextClause object at 0x7fe9935787f0>, for_update=False)}, False, True)]] is not false : Database schema and SQLAlchemy model are not in sync: [[('modify_nullable', None, 'task_fail', 'start_date', {'existing_type': TIMESTAMP(fsp=6), 'existing_server_default': DefaultClause(<sqlalchemy.sql.elements.TextClause object at 0x7fe9935787f0>, for_update=False)}, False, True)]][0m
```

**Comment by ashb** [comment]
That travis error is saying that the result of the migrations is not the same as the definition of the model.

**Comment by dossett** [comment]
Is the model defined in code somewhere?  Sorry, this is a new part of the code for me.

**Comment by ashb** [comment]
It is - in `airflow/models/*.py` - but that needs to be "DB-server" agnostic, so can't be updated to have a mysql-specific default.

I'm seeing if I can find a way of changing the default_value of the model based on the current engine.

The other option may be to allow those specific differences for mysql in the tests.

File: airflow/migrations/versions/0e2a74e0fc9f_add_time_zone_awareness.py [hunk 1]
```diff
@@ -153,13 +151,22 @@ def upgrade():
...
+            server_default=sa.text('CURRENT_TIMESTAMP(6)')
         )
         op.alter_column(
             table_name="xcom",
             column_name="execution_date",
             type_=mysql.TIMESTAMP(fsp=6),
+            nullable=False,
+            server_default=sa.text('CURRENT_TIMESTAMP(6)')
         )
+        conn.execute("alter table task_instance alter column execution_date drop default")
```
Comment (ashb) [inline airflow/migrations/versions/0e2a74e0fc9f_add_time_zone_awareness.py:167]: Doesn't this undo the change made on L135?

File: airflow/migrations/versions/0a2a5b66e19d_add_task_reschedule_table.py [hunk 2]
```diff
@@ -66,13 +66,14 @@ def upgrade():
...
         sa.Column('dag_id', sa.String(length=250), nullable=False),
-        # use explicit server_default=None otherwise mysql implies defaults for first timestamp column
-        sa.Column('execution_date', timestamp(), nullable=False, server_default=None),
+        sa.Column('execution_date', timestamp(), nullable=False,
+                  server_default=sa.text('CURRENT_TIMESTAMP(6)')),
         sa.Column('try_number', sa.Integer(), nullable=False),
-        sa.Column('start_date', timestamp(), nullable=False),
-        sa.Column('end_date', timestamp(), nullable=False),
+        sa.Column('start_date', timestamp(), nullable=False, server_default=sa.text('CURRENT_TIMESTAMP(6)')),
+        sa.Column('end_date', timestamp(), nullable=False, server_default=sa.text('CURRENT_TIMESTAMP(6)')),
```
Comment (ashb) [inline airflow/migrations/versions/0a2a5b66e19d_add_task_reschedule_table.py:73]: Additionally: I am not sure this is the right fix: we don't want the default value of these columns to be "Now()" or any variation of it.

Comment (dossett) [reply airflow/migrations/versions/0a2a5b66e19d_add_task_reschedule_table.py:73]: I'm not a Mysql expert by any means, but I understand the issue to be that if you specify a timestamp to column to be not null you have to supply a default value.  Given the model also knows that it's not null it *should* be the case that we never have to rely on the database's default?

Is the current_timestamp better or worse that any other default that could be used at the database level?

Maybe another option is to require mysql to use this flag? https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_explicit_defaults_for_timestamp

No idea how feasible that would be.

**Comment by ashb** [comment]
Requiring that flag is what the migration currently does, but unfortunately that means it can't work on Google's CloudDB-Mysql (that can't be set.)

It's been so long since I used MySQL I don't know either.

Could we use DATETIME instead of TIMESTAMP columns instead? That would (I think?) side-step this entire issue of odd behaviour on TIMESTAMP columns, right?

**Comment by dossett** [comment]
replying to keep this PR alive, I hope to follow-up soon

**Comment by dephusluke** [comment]
@dossett is it recommended not to use mysql flavour of CloudSQL due to this issue?

**Comment by ashb** [comment]
@dephusluke I personally would recommend you use Postgres where possible - MySQL has historically had dead-lock issues in some cases on Airflow (the ones we know about solved, but there may be more) and generally I have found it's query planner and optimizer to be sub-par compared to Postgres.

And given Google CloudSQL supports PostgresSQL I would point you towards that.
